### PR TITLE
Add multitude of features to the toolkit

### LINKED ISSUE #89: Add many new features to LBT_Toolkit
#### Description:

Lots of great work needs to be merged. Let do this team!

### PR DESCRIPTION
**Pull request by tg359**
### Issues addressed by this PR

Closes #89

@tg359 to add list of new features here

### Depends on
Build this first before testing:
BHoM/Python_Toolkit#96

### CODE REVIEW

File: LadybugTools_Engine/Create/Shelter.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+using BH.oM.Dimensional;
+using BH.oM.Analytical.Elements;
+
+namespace BH.Engine.LadybugTools
+{
+    public static partial class Create
+    {
+        [Description("Create a Shelter from geometry.")]
+        [Input("polyline", "A planar Polyline.")]
+        [Input("windPorosity", "The porosity of this shelter object to wind.")]
```
Comment (jamesramsden-bh) [inline LadybugTools_Engine/Create/Shelter.cs:39]: Requires range check between 0 to 1

File: LadybugTools_Engine/Create/Shelter.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        [Input("radiationPorosity", "The porosity of this shelter object to radiation.")]
```
Comment (jamesramsden-bh) [inline LadybugTools_Engine/Create/Shelter.cs:40]: Requires range check between 0 and 1

File: LadybugTools_Engine/Create/Shelter.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+                BH.Engine.Base.Compute.RecordError("Input object is not closed.");
+            }
+
+            List<List<double>> vertices = new List<List<double>>();
+            for (int i = 0; i < polyline.ControlPoints.Count - 1; i++)
+            {
+                vertices.Add(new List<double>() { polyline.ControlPoints[i].X, polyline.ControlPoints[i].Y, polyline.ControlPoints[i].Z });
+            }
+
+            return new Shelter() { Vertices = vertices, WindPorosity = windPorosity, RadiationPorosity = radiationPorosity };
```
Comment (jamesramsden-bh) [inline LadybugTools_Engine/Create/Shelter.cs:60]: There are more BHoM-friendly ways of defining the control points of the shelter - a `List<Point>` - or even saving a Polyline object to the Shelter object would be more appropriate.

**Review by jamesramsden-bh — CHANGES_REQUESTED** [review]
Only a partial review done so far - a few comments made.

File: LadybugTools_Engine/Create/Shelter.cs (hunk 3 above)
Comment (jamesramsden-bh) [reply LadybugTools_Engine/Create/Shelter.cs:60]: After reviewing this in more detail, the original solution is necessary to allow for the correct serialisation/deserialisation.

File: LadybugTools_Engine/Create/Shelter.cs (hunk 1 above)
Comment (jamesramsden-bh) [reply LadybugTools_Engine/Create/Shelter.cs]: Resolved via Create.Shelter

File: LadybugTools_Engine/Create/Shelter.cs (hunk 2 above)
Comment (jamesramsden-bh) [reply LadybugTools_Engine/Create/Shelter.cs]: Resolved via Create.Shelter

**Comment by jamesramsden-bh** [comment]
@BHoMBot check required

**Review by jamesramsden-bh — DISMISSED** [review]
Tested BHoM methods in GH with a range of shelters and typologies, and exploded outputs to ensure all looks sensible. Ran unit tests on Python stuff.

**Comment by jamesramsden-bh** [comment]
@BHoMBot check code-compliance

**Comment by jamesramsden-bh** [comment]
@BHoMBot check core

**Comment by jamesramsden-bh** [comment]
@BHoMBot check core

**Comment by jamesramsden-bh** [comment]
@BHoMBot check copyright-compliance
@BHoMBot check dataset-compliance
@BHoMBot check documentation-compliance
@BHoMBot check null-handling
@BHoMBot check project-compliance
@BHoMBot check unit-tests

**Comment by jamesramsden-bh** [comment]
@BHoMBot check null-handling

**Comment by jamesramsden-bh** [comment]
@BHoMBot check null-handling

**Comment by jamesramsden-bh** [comment]
@bhom check null-handling

**Comment by jamesramsden-bh** [comment]
@BHoMBot check null-handling

**Comment by jamesramsden-bh** [comment]
@BHoMBot check null-handling

File: LadybugTools_Engine/Query/GetMaterial.cs [hunk 4]
```diff
@@ -20,237 +20,48 @@
...
-                        Thickness = 0.0007999999999999979,
-                        Conductivity = 45.24971874361766,
-                        Density = 7824.017889489713,
-                        SpecificHeat = 499.67760800730963,
-                        Roughness = Roughness.Smooth,
-                        ThermalAbsorptance = 0.9,
-                        SolarAbsorptance = 0.7,
-                        VisibleAbsorptance = 0.7,
-                    };
+            if (filter.IsNullOrEmpty())
```
Comment (FraserGreenroyd) [inline LadybugTools_Engine/Query/GetMaterial.cs:43]:
```suggestion
            if (string.IsNullOrEmpty(filter))
```

File: LadybugTools_Engine/Query/GetTypology.cs [hunk 5]
```diff
@@ -20,591 +20,49 @@
...
+        [Description("Returns a list of Typology objects from the Python predefined Typologies list.")]
+        [Input("filter", "Text to filter the resultant list by. Filter applies to the typology name. Leave blank to return all typologies.")]
+        [Output("typologies", "A list of Typology objects.")]
+        public static List<Typology> GetTypology(string filter = "")
         {
-            if (typology == Typologies.Undefined)
-            {
-                BH.Engine.Base.Compute.RecordError("A pre-defined typology must be passed in order to return an object.");
-            }
+            if (filter.IsNullOrEmpty())
```
Comment (FraserGreenroyd) [inline LadybugTools_Engine/Query/GetTypology.cs:42]:
```suggestion
            if (string.IsNullOrEmpty(filter))
```

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]

**Comment by jamesramsden-bh** [comment]
@BHoMBot check null-handling

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance

File: LadybugTools_Engine/Query/GetMaterial.cs [hunk 6]
```diff
@@ -20,237 +20,48 @@
...
     public static partial class Query
     {
-        [Description("Get a material object from its Enum.")]
-        [Input("material", "An enum value representing a pre-defined ILBTMaterial object.")]
-        [Output("material", "A material object to pass into the External Comfort workflow.")]
-        public static ILBTMaterial GetMaterial(this Materials material)
+        [Description("Returns a list of materials from the Python Materials list.")]
+        [Input("filter", "Text to filter the resultant list by. Filter applies to the material identifier. Leave blank to return all materials.")]
+        [Output("materials", "A list of materials.")]
+        public static List<ILBTMaterial> GetMaterial(string filter = "")
```
Comment (FraserGreenroyd) [inline LadybugTools_Engine/Query/GetMaterial.cs:41]:
```suggestion
        [PreviousVersion("6.1", "BH.Engine.LadybugTools.Query.GetMaterial(BH.oM.LadybugTools.Materials)")]
        public static List<ILBTMaterial> GetMaterial(string filter = "")
```

File: LadybugTools_Engine/Query/GetTypology.cs [hunk 7]
```diff
@@ -20,591 +20,49 @@
...
     public static partial class Query
     {
-        [Description("Get a material object from it's Enum.")]
-        [Input("typology", "An enum value representing a pre-defined Typology object.")]
-        [Output("typology", "A Typology object to pass into the External Comfort workflow.")]
-        public static Typology GetTypology(this Typologies typology)
+        [Description("Returns a list of Typology objects from the Python predefined Typologies list.")]
+        [Input("filter", "Text to filter the resultant list by. Filter applies to the typology name. Leave blank to return all typologies.")]
+        [Output("typologies", "A list of Typology objects.")]
+        public static List<Typology> GetTypology(string filter = "")
```
Comment (FraserGreenroyd) [inline LadybugTools_Engine/Query/GetTypology.cs:40]:
```suggestion
        [PreviousVersion("6.1", "BH.Engine.LadybugTools.Query.GetTypology(BH.oM.LadybugTools.Typologies)")]
        public static List<Typology> GetTypology(string filter = "")
```

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]

**Comment by jamesramsden-bh** [comment]
@BHoMBot check versioning

**Comment by jamesramsden-bh** [comment]
@BHoMBot check copyright-compliance
@BHoMBot check dataset-compliance
@BHoMBot check documentation-compliance
@BHoMBot check null-handling
@BHoMBot check project-compliance
@BHoMBot check unit-tests
@BHoMBot check core
@BHoMBot check installer
@BHoMBot check code-compliance

**Comment by jamesramsden-bh** [comment]
@BHoMBot check serialisation

**Review by jamesramsden-bh — APPROVED** [review]
Approved - null check has been dispensated.

**Review by jamesramsden-bh — CHANGES_REQUESTED** [review]
Retracting my pass - because the unit tests have failed! And on the same methods that were causing null check trouble. To investigate next week.

**Comment by jamesramsden-bh** [comment]
Owing to time pressures and fixes required, this PR will be closed for now. This commit will be revisited in the next milestone.

**Comment by jamesramsden-bh** [comment]
This PR has been re-opened in an attempt to get it merged. Tasks to get it merged:

 - [ ] Get automated tests on Python side passing
 - [ ] Fix any existing BHoM methods that can be easily fixed (usually namespace changes after Python refactoring)
 - [ ] Check and fix any versioning issues

**Comment by FraserGreenroyd** [comment]
I would also recommend a rebase @jamesramsden-bh @tg359

File: LadybugTools_Engine/Query/GetMaterial.cs (hunk 6 above)
Comment (FraserGreenroyd) [reply LadybugTools_Engine/Query/GetMaterial.cs:42]: Needs updating to 6.2 now 😄

File: LadybugTools_Engine/Query/GetTypology.cs (hunk 7 above)
Comment (FraserGreenroyd) [reply LadybugTools_Engine/Query/GetTypology.cs:43]: Same here, update to 6.2

File: LadybugTools_Engine/Compute/EPWtoCSV.cs [hunk 8]
```diff
@@ -33,8 +34,9 @@ public static partial class Compute
     {
         [Description("Convert an EPW file into a CSV and return the path to that CSV.")]
         [Input("epwFile", "An EPW file.")]
+        [Input("includeAdditional", "Add sun position and psychrometric properties to the resultant CSV.")]
         [Output("csv", "The generated CSV file.")]
-        public static string EPWtoCSV(string epwFile)
+        public static string EPWtoCSV(string epwFile, bool includeAdditional = false)
```
Comment (FraserGreenroyd) [inline LadybugTools_Engine/Compute/EPWtoCSV.cs:39]:
```suggestion
        [PreviousVersion("6.2", "BH.Engine.LadybugTools.Compute.EPWtoCSV(System.String)")]
        public static string EPWtoCSV(string epwFile, bool includeAdditional = false)
```

**Review by jamesramsden-bh — DISMISSED** [review]
Reinstalled the venv and ran the test script, all working fine. Have also been working directly with the Python via my own Jupyter scripts without issues. 

Automated tests also fully pass.

**Comment by jamesramsden-bh** [comment]
@BHoMBot check required

**Comment by jamesramsden-bh** [comment]
@BHoMBot check project-compliance

**Comment by jamesramsden-bh** [comment]
@BHoMBot fix project file ref. 14069254324

**Comment by tg359** [comment]
@BHoMBot fix project file ref. 14069254324

**Comment by jamesramsden-bh** [comment]
@BHoMBot check versioning

**Comment by jamesramsden-bh** [comment]
@BHoMBot check versioning

**Comment by FraserGreenroyd** [comment]
@BHoMBot check versioning

**Comment by FraserGreenroyd** [comment]
@BHoMBot check versioning

**Comment by FraserGreenroyd** [comment]
@BHoMBot check versioning

**Comment by jamesramsden-bh** [comment]
@BHoMBot check versioning

**Comment by jamesramsden-bh** [comment]
@BHoMBot check versioning

**Comment by jamesramsden-bh** [comment]
@BHoMBot check unit-tests
@BHoMBot check serialisation
@BHoMBot check project-compliance
@BHoMBot check null-handling
@BHoMBot check documentation-compliance
@BHoMBot check dataset-compliance
@BHoMBot check core
@BHoMBot check copyright-compliance
@BHoMBot check code-compliance

**Comment by jamesramsden-bh** [comment]
@BHoMBot check installer

**Review by FraserGreenroyd — APPROVED** [review]
I've reviewed the code through with @jamesramsden-bh and am happy my changes have been addressed and this work is working suitably for the work done by @tg359 and @jamesramsden-bh so happy for this to be merged and deployed in 6.2

**Comment by jamesramsden-bh** [comment]
@BHoMBot check ready-to-merge

**Pull request merged**